src/txreconciliation.cpp

```diff
@@ -0,0 +1,996 @@
+    }
+
+    bool HandleInitialSketch(std::unordered_map<NodeId, ReconciliationState>::iterator& recon_state,
+        const std::vector<uint8_t>& skdata,
+        // returning values
+        std::vector<uint32_t>& txs_to_request, std::vector<uint256>& txs_to_announce, std::optional<bool>& result)
+    {
+        assert(recon_state->second.m_we_initiate);
+        assert(recon_state->second.m_state_init_by_us.m_phase == Phase::INIT_REQUESTED);
+
+        uint32_t remote_sketch_capacity = uint32_t(skdata.size() / BYTES_PER_SKETCH_CAPACITY);
+        // Protocol violation: our peer exceeded the sketch capacity, or sent a malformed sketch.
+        if (remote_sketch_capacity > MAX_SKETCH_CAPACITY) {
+            return false;
+        }
+
+        Minisketch local_sketch = recon_state->second.ComputeBaseSketch(remote_sketch_capacity);
+        Minisketch remote_sketch;
+        if (remote_sketch_capacity != 0) {
+            remote_sketch = Minisketch(RECON_FIELD_SIZE, 0, remote_sketch_capacity).Deserialize(skdata);
+        }
+
+        // Remote sketch is empty in two cases per which reconciliation is pointless:
+        // 1. the peer has no transactions for us
+        // 2. we told the peer we have no transactions for them while initiating reconciliation.
+        // In case (2), local sketch is also empty.
+        if (remote_sketch_capacity == 0 || !remote_sketch || !local_sketch) {
+
+            // Announce all transactions we have.
+            txs_to_announce = recon_state->second.m_local_set.GetAllTransactions();
+
+            // Update local reconciliation state for the peer.
+            recon_state->second.FinalizeInitByUs(true, DEFAULT_RECON_Q);
+            recon_state->second.m_state_init_by_us.m_phase = Phase::NONE;
+
+            result = false;
+
+            LogPrint(BCLog::NET, "Reconciliation we initiated with peer=%d terminated due to empty sketch. " /* Continued */
+                "Announcing all %i transactions from the local set.\n", recon_state->first, txs_to_announce.size());
+
+            return true;
+        }
+
+        assert(remote_sketch);
+        assert(local_sketch);
+        // Attempt to decode the set difference
+        size_t max_elements = minisketch_compute_max_elements(RECON_FIELD_SIZE, remote_sketch_capacity, RECON_FALSE_POSITIVE_COEF);
+        std::vector<uint64_t> differences(max_elements);
+        if (local_sketch.Merge(remote_sketch).Decode(differences)) {
+            // Initial reconciliation step succeeded.
+
+            // Identify locally/remotely missing transactions.
+            recon_state->second.m_local_set.GetRelevantIDsFromShortIDs(differences, txs_to_request, txs_to_announce);
+
+            // Update local reconciliation state for the peer.
+            size_t local_set_size = recon_state->second.m_local_set.GetSize();
+            recon_state->second.FinalizeInitByUs(true,
+                RecomputeQ(local_set_size, txs_to_request.size(), txs_to_announce.size()));
```

Comment:
Note, coefficient recomputation doesn't seem covered by unit tests but sounds to be by functional tests. I would favor covering it unit tests too. An unrelated change in `p2p_erlay.py` in the future might remove coverage, up to you.